modules/gapi/test/common/gapi_core_tests_inl.hpp

```diff
@@ -1267,6 +1267,39 @@ TEST_P(NormalizeTest, Test)
     }
 }
 
+TEST_P(InitOutTest, TestWithAdd)
+{
+    in_mat1 = cv::Mat(sz, type);
+    in_mat2 = cv::Mat(sz, type);
+    cv::randu(in_mat1, cv::Scalar::all(0), cv::Scalar::all(100));
+    cv::randu(in_mat2, cv::Scalar::all(0), cv::Scalar::all(100));
+
+    // G-API code //////////////////////////////////////////////////////////////
+    cv::GMat in1, in2, out;
+    out = cv::gapi::add(in1, in2, dtype);
+    cv::GComputation c(cv::GIn(in1, in2), cv::GOut(out));
+
+    const auto run_and_compare = [&c, this] ()
+    {
+        // G-API code //////////////////////////////////////////////////////////////
+        c.apply(cv::gin(in_mat1, in_mat2), cv::gout(out_mat_gapi), getCompileArgs());
+
+        // OpenCV code /////////////////////////////////////////////////////////////
+        cv::add(in_mat1, in_mat2, out_mat_ocv, cv::noArray());
+
+        // Comparison //////////////////////////////////////////////////////////////
+        EXPECT_EQ(0, cv::countNonZero(out_mat_gapi != out_mat_ocv));
+        EXPECT_EQ(out_mat_gapi.size(), sz);
+    };
+
+    // run for uninitialized output
+    run_and_compare();
+
+    // run for initialized output (can be initialized with a different size)
+    initOutMats(out_sz, type);
+    run_and_compare();
+}
```

Comment:
This test is more a smoke test, making sure a thing "just works", but is not an API contract test (demonstrating the behavior we commit on).

Please cover the following situations explicitly:
1. Output matrix is empty. Check if (a) G-API can handle that and (b) the result is correct.
2. Output matrix has exact (correct) meta (type and format). Check that after G-API is finished, data wasn't reallocated.
3. Output matrix has a different (than expected) meta but its physical buffer is enough to hold the result. Check that after G-API is finished, the meta is set correctly but reallocation didn't happen.
4. Output matrix has a correct type but a smaller size (less than required). Check that after G-API is finished, data is reallocated.
5. Output matrix has a correct type and a larger size. Check that after G-API is finished, Mat is set to the right size but reallocation didn't happen.
6. An interesting case when output mat is a view to another mat. Check which cases of the above apply to this case.

Reply:
(addressed in the another pr #15012